test/threads_exec.jl

```diff
@@ -1312,4 +1313,80 @@ end
+            t = Threads.@spawn peakflops()
+            wait(t)
+            end_time = time_ns()
+            wall_time_delta = end_time - start_time
+            @test t.is_timing_enabled
+            @test Base.task_cpu_time_ns(t) > 0
+            @test Base.task_wall_time_ns(t) > 0
+            @test Base.task_wall_time_ns(t) >= Base.task_cpu_time_ns(t)
+            @test wall_time_delta > Base.task_wall_time_ns(t)
+        finally
+            Base.task_timing(false)
+        end
+    end
+    @testset "disabled" begin
+        t = Threads.@spawn peakflops()
+        wait(t)
+        @test !t.is_timing_enabled
+        @test Base.task_cpu_time_ns(t) == 0
+        @test Base.task_wall_time_ns(t) == 0
+    end
+end
+
+@testset "task time counters: lots of spawns" begin
+    using Dates
+    try
+        Base.task_timing(true)
+        # create more tasks than we have threads.
+        # - all tasks must have: cpu time <= wall time
+        # - some tasks must have: cpu time < wall time
+        # - summing across all tasks we must have: total cpu time <= available cpu time
+        n_tasks = 2 * Threads.nthreads(:default)
+        cpu_times = Vector{UInt64}(undef, n_tasks)
+        wall_times = Vector{UInt64}(undef, n_tasks)
+        start_time = time_ns()
+        @sync begin
+            for i in 1:n_tasks
+                start_time_i = time_ns()
+                task_i = Threads.@spawn peakflops()
+                Threads.@spawn begin
+                    wait(task_i)
+                    end_time_i = time_ns()
+                    wall_time_delta_i = end_time_i - start_time_i
+                    cpu_times[$i] = cpu_time_i = Base.task_cpu_time_ns(task_i)
+                    wall_times[$i] = wall_time_i = Base.task_wall_time_ns(task_i)
+                    # task should have recorded some cpu-time and some wall-time
+                    @test cpu_time_i > 0
+                    @test wall_time_i > 0
+                    # task cpu-time cannot be greater than its wall-time
+                    @test wall_time_i >= cpu_time_i
+                    # task wall-time must be less than our manually measured wall-time
+                    # between calling `@spawn` and returning from `wait`.
+                    @test wall_time_delta_i > wall_time_i
+                end
+            end
+        end
+        end_time = time_ns()
+        wall_time_delta = (end_time - start_time)
+        available_cpu_time = wall_time_delta * Threads.nthreads(:default)
+        summed_cpu_time = sum(cpu_times)
+        # total CPU time from all tasks can't exceed what was actually available.
+        @test available_cpu_time > summed_cpu_time
+        # some tasks must have cpu-time less than their wall-time, because we had more tasks
+        # than threads.
+        summed_wall_time = sum(wall_times)
+        @test summed_wall_time > summed_cpu_time
+    finally
+        Base.task_timing(false)
+    end
+end
+
 end # main testset
```

Comment:
Maybe you could add a bit more testing around edge cases, like if you create a Task but don't run it, what its numbers look like, or if you create a task and try to run it, but you provide a bad value for the function so it never actually runs, what happens to the numbers.

And what happens if a task throws an exception, do its numbers still get cleaned up correctly (they don't keep incrementing), etc.

Reply:
note to self: still need to add these tests

Reply:
okay, added more tests -- please take a look (more suggestions welcome!)